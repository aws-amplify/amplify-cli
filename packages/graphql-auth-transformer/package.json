{
  "name": "graphql-auth-transformer",
  "version": "6.18.5",
  "description": "Implements the @auth directive for the appsync model transform.",
  "repository": {
    "type": "git",
    "url": "https://github.com/aws-amplify/amplify-cli.git",
    "directory": "packages/graphql-auth-transformer"
  },
  "author": "Amazon Web Services",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "keywords": [
    "graphql",
    "appsync",
    "aws"
  ],
  "scripts": {
    "test": "jest",
    "test-ci": "jest --ci -i",
    "build": "tsc",
    "clean": "rimraf ./lib",
    "watch": "tsc -w"
  },
  "dependencies": {
    "graphql": "^14.5.8",
    "graphql-connection-transformer": "4.18.4",
    "graphql-mapping-template": "4.13.4",
    "graphql-transformer-common": "4.17.4",
    "graphql-transformer-core": "6.19.4"
  },
  "devDependencies": {
    "@types/node": "^10.17.13",
    "cloudform-types": "^4.2.0",
    "graphql-dynamodb-transformer": "6.19.5",
    "graphql-elasticsearch-transformer": "4.7.7",
    "graphql-function-transformer": "2.3.12",
    "rimraf": "^3.0.0"
  },
  "jest": {
    "testURL": "http://localhost",
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "testRegex": "(src/__tests__/.*.test.ts)$",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ],
    "collectCoverage": true
  }
}
