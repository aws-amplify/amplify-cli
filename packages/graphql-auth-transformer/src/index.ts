import { ModelAuthTransformer } from './ModelAuthTransformer';
export * from './ModelAuthTransformer';
export default ModelAuthTransformer;
