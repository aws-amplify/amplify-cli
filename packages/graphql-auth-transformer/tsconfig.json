{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "strict": false, // TODO enable
    "rootDir": "src",
    "outDir": "lib"
  },
  "references": [
    {"path": "../graphql-connection-transformer"},
    {"path": "../graphql-dynamodb-transformer"},
    {"path": "../graphql-elasticsearch-transformer"},
    {"path": "../graphql-mapping-template"},
    {"path": "../graphql-transformer-common"},
    {"path": "../graphql-transformer-core"}
  ]
}
