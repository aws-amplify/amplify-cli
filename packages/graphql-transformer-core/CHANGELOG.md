# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="1.0.28"></a>
## [1.0.28](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.28-beta.0...graphql-transformer-core@1.0.28) (2018-10-18)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.28-beta.0"></a>
## [1.0.28-beta.0](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.12...graphql-transformer-core@1.0.28-beta.0) (2018-10-12)


### Bug Fixes

* **graphql-transformer-core:** Fix Readme.md docs for the `[@auth](https://github.com/auth)` directive ([db6ff7a](https://github.com/aws-amplify/amplify-cli/commit/db6ff7a))




<a name="1.0.12"></a>
## [1.0.12](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.11...graphql-transformer-core@1.0.12) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.11"></a>
## [1.0.11](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.10...graphql-transformer-core@1.0.11) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.10"></a>
## [1.0.10](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.9...graphql-transformer-core@1.0.10) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.9"></a>
## [1.0.9](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.8...graphql-transformer-core@1.0.9) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.8"></a>
## [1.0.8](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.7...graphql-transformer-core@1.0.8) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.7"></a>
## [1.0.7](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.6...graphql-transformer-core@1.0.7) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.6"></a>
## [1.0.6](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.5...graphql-transformer-core@1.0.6) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.5"></a>
## [1.0.5](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.4...graphql-transformer-core@1.0.5) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.4"></a>
## 1.0.4 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.3"></a>
## 1.0.3 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.2"></a>
## 1.0.2 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.1"></a>
## 1.0.1 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core
