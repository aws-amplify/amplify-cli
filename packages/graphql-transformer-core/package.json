{
  "name": "graphql-transformer-core",
  "version": "6.19.4",
  "description": "A framework to transform from GraphQL SDL to AWS CloudFormation.",
  "repository": {
    "type": "git",
    "url": "https://github.com/aws-amplify/amplify-cli.git",
    "directory": "packages/graphql-transformer-core"
  },
  "author": "Amazon Web Services",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "keywords": [
    "graphql",
    "cloudformation",
    "aws"
  ],
  "scripts": {
    "test": "jest",
    "test-ci": "jest --ci -i",
    "build": "tsc",
    "watch": "tsc -w",
    "clean": "rimraf ./lib"
  },
  "dependencies": {
    "cloudform-types": "^4.2.0",
    "deep-diff": "^1.0.2",
    "fs-extra": "^8.1.0",
    "glob": "^7.1.6",
    "graphql": "^14.5.8",
    "graphql-transformer-common": "4.17.4"
  },
  "devDependencies": {
    "@types/fs-extra": "^8.0.1",
    "@types/node": "^10.17.13"
  },
  "jest": {
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "testURL": "http://localhost",
    "testRegex": "(src/__tests__/.*|(\\.|/)(test|spec))\\.(jsx?|tsx?)$",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ],
    "collectCoverage": true
  }
}
