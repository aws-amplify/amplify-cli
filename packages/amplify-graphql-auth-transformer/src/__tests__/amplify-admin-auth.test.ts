import { AuthTransformer } from '../graphql-auth-transformer';
import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import { GraphQLTransform } from '@aws-amplify/graphql-transformer-core';
import _ from 'lodash';

const ADMIN_UI_ROLES = ['us-fake-1_uuid_Full-access/CognitoIdentityCredentials', 'us-fake-1_uuid_Manage-only/CognitoIdentityCredentials'];

test('test simple model with public auth rule and amplify admin app is present', () => {
  const validSchema = `
  type Post @model @auth(rules: [{allow: public}]) {
    id: ID!
    title: String!
    createdAt: String
    updatedAt: String
  }`;
  const transformer = new GraphQLTransform({
    authConfig: {
      defaultAuthentication: {
        authenticationType: 'API_KEY',
      },
      additionalAuthenticationProviders: [
        {
          authenticationType: 'AWS_IAM',
        },
      ],
    },
    transformers: [
      new ModelTransformer(),
      new AuthTransformer({
        adminRoles: ADMIN_UI_ROLES,
      }),
    ],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.schema).toContain('Post @aws_api_key @aws_iam');
});

test('Test simple model with public auth rule and amplify admin app is not enabled', () => {
  const validSchema = `
      type Post @model @auth(rules: [{allow: public}]) {
          id: ID!
          title: String!
          createdAt: String
          updatedAt: String
      }
      `;
  const transformer = new GraphQLTransform({
    authConfig: {
      defaultAuthentication: {
        authenticationType: 'API_KEY',
      },
      additionalAuthenticationProviders: [],
    },
    transformers: [new ModelTransformer(), new AuthTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.schema).not.toContain('Post @aws_api_key @aws_iam');
});

test('Test model with public auth rule without all operations and amplify admin app is present', () => {
  const validSchema = `
      type Post @model @auth(rules: [{allow: public, operations: [read, update]}]) {
          id: ID!
          title: String!
          createdAt: String
          updatedAt: String
      }
      `;
  const transformer = new GraphQLTransform({
    authConfig: {
      defaultAuthentication: {
        authenticationType: 'API_KEY',
      },
      additionalAuthenticationProviders: [
        {
          authenticationType: 'AWS_IAM',
        },
      ],
    },
    transformers: [
      new ModelTransformer(),
      new AuthTransformer({
        adminRoles: ADMIN_UI_ROLES,
      }),
    ],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();

  expect(out.schema).toContain('type Post @aws_iam @aws_api_key');
  expect(out.schema).toContain('createPost(input: CreatePostInput!, condition: ModelPostConditionInput): Post @aws_api_key @aws_iam');
  expect(out.schema).toContain('updatePost(input: UpdatePostInput!, condition: ModelPostConditionInput): Post @aws_api_key @aws_iam');
  expect(out.schema).toContain('deletePost(input: DeletePostInput!, condition: ModelPostConditionInput): Post @aws_api_key @aws_iam');

  // No Resource extending Auth and UnAuth role
  const policyResources = Object.values(out.rootStack.Resources!).filter(r => r.Type === 'AWS::IAM::ManagedPolicy');
  expect(policyResources).toHaveLength(0);
});

test('Test simple model with private auth rule and amplify admin app is present', () => {
  const validSchema = `
      type Post @model @auth(rules: [{allow: groups, groups: ["Admin", "Dev"]}]) {
          id: ID!
          title: String!
          createdAt: String
          updatedAt: String
      }
      `;
  const transformer = new GraphQLTransform({
    authConfig: {
      defaultAuthentication: {
        authenticationType: 'AMAZON_COGNITO_USER_POOLS',
      },
      additionalAuthenticationProviders: [
        {
          authenticationType: 'AWS_IAM',
        },
      ],
    },
    transformers: [
      new ModelTransformer(),
      new AuthTransformer({
        adminRoles: ADMIN_UI_ROLES,
      }),
    ],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.schema).toContain('type Post @aws_iam @aws_cognito_user_pools');
});

test('Test simple model with private auth rule and amplify admin app not enabled', () => {
  const validSchema = `
      type Post @model @auth(rules: [{allow: groups, groups: ["Admin", "Dev"]}]) {
          id: ID!
          title: String!
          createdAt: String
          updatedAt: String
      }
      `;
  const transformer = new GraphQLTransform({
    authConfig: {
      defaultAuthentication: {
        authenticationType: 'AMAZON_COGNITO_USER_POOLS',
      },
      additionalAuthenticationProviders: [
        {
          authenticationType: 'AWS_IAM',
        },
      ],
    },
    transformers: [new ModelTransformer(), new AuthTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.schema).not.toContain('type Post @aws_iam @aws_cognito_user_pools');
});

test('Test simple model with private auth rule, few operations, and amplify admin app enabled', () => {
  const validSchema = `
      type Post @model @auth(rules: [{allow: groups, groups: ["Admin", "Dev"], operations: [read]}]) {
          id: ID!
          title: String!
          createdAt: String
          updatedAt: String
      }
      `;
  const transformer = new GraphQLTransform({
    authConfig: {
      defaultAuthentication: {
        authenticationType: 'AMAZON_COGNITO_USER_POOLS',
      },
      additionalAuthenticationProviders: [
        {
          authenticationType: 'AWS_IAM',
        },
      ],
    },
    transformers: [
      new ModelTransformer(),
      new AuthTransformer({
        adminRoles: ADMIN_UI_ROLES,
      }),
    ],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.schema).toContain('type Post @aws_iam @aws_cognito_user_pools');
  expect(out.schema).toContain(
    'createPost(input: CreatePostInput!, condition: ModelPostConditionInput): Post @aws_iam @aws_cognito_user_pools',
  );
  expect(out.schema).toContain(
    'updatePost(input: UpdatePostInput!, condition: ModelPostConditionInput): Post @aws_iam @aws_cognito_user_pools',
  );
  expect(out.schema).toContain(
    'deletePost(input: DeletePostInput!, condition: ModelPostConditionInput): Post @aws_iam @aws_cognito_user_pools',
  );

  // No Resource extending Auth and UnAuth role
  const policyResources = Object.values(out.rootStack.Resources!).filter(r => r.Type === 'AWS::IAM::ManagedPolicy');
  expect(policyResources).toHaveLength(0);
});

test('Test simple model with private IAM auth rule, few operations, and amplify admin app is not enabled', () => {
  const validSchema = `
      type Post @model @auth(rules: [{allow: private, provider: iam, operations: [read, update]}]) {
          id: ID!
          title: String!
          createdAt: String
          updatedAt: String
      }
      `;
  const transformer = new GraphQLTransform({
    authConfig: {
      defaultAuthentication: {
        authenticationType: 'AMAZON_COGNITO_USER_POOLS',
      },
      additionalAuthenticationProviders: [
        {
          authenticationType: 'AWS_IAM',
        },
      ],
    },
    transformers: [
      new ModelTransformer(), 
      new AuthTransformer( {
        identityPoolId: 'testIdentityPoolId',
      })
    ],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.schema).toContain('Post @aws_iam');
  expect(out.schema).not.toContain(
    'createPost(input: CreatePostInput!, condition: ModelPostConditionInput): Post @aws_iam @aws_cognito_user_pools',
  );
  expect(out.schema).not.toContain('deletePost(input: DeletePostInput!): Post @aws_iam');
  expect(out.schema).not.toContain('updatePost(input: UpdatePostInput!): Post @aws_iam');

  expect(out.schema).toContain('getPost(id: ID!): Post @aws_iam');
  expect(out.schema).toContain('listPosts(filter: ModelPostFilterInput, limit: Int, nextToken: String): ModelPostConnection @aws_iam');

  expect(out.resolvers['Mutation.updatePost.auth.1.res.vtl']).toMatchSnapshot();
  expect(out.resolvers['Mutation.updatePost.auth.1.res.vtl']).toContain(`#if( ($ctx.identity.userArn == $ctx.stash.authRole) || ($ctx.identity.cognitoIdentityPoolId == "testIdentityPoolId" && $ctx.identity.cognitoIdentityAuthType == "authenticated") )`);
});

test('Test simple model with AdminUI enabled should add IAM policy only for fields that have explicit IAM auth', () => {
  const validSchema = `
      type Post @model @auth(rules: [{allow: private, provider: iam, operations: [read]}]) {
          id: ID!
          title: String!
          createdAt: String
          updatedAt: String
      }
      `;
  const transformer = new GraphQLTransform({
    authConfig: {
      defaultAuthentication: {
        authenticationType: 'AMAZON_COGNITO_USER_POOLS',
      },
      additionalAuthenticationProviders: [
        {
          authenticationType: 'AWS_IAM',
        },
      ],
    },
    transformers: [
      new ModelTransformer(),
      new AuthTransformer({
        adminRoles: ADMIN_UI_ROLES,
      }),
    ],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.schema).toContain('Post @aws_iam @aws_cognito_user_pool');
  expect(out.schema).toContain(
    'createPost(input: CreatePostInput!, condition: ModelPostConditionInput): Post @aws_iam @aws_cognito_user_pools',
  );
  expect(out.schema).toContain(
    'updatePost(input: UpdatePostInput!, condition: ModelPostConditionInput): Post @aws_iam @aws_cognito_user_pools',
  );
  expect(out.schema).toContain(
    'deletePost(input: DeletePostInput!, condition: ModelPostConditionInput): Post @aws_iam @aws_cognito_user_pools',
  );

  expect(out.schema).toContain('getPost(id: ID!): Post @aws_iam');
  expect(out.schema).toContain('listPosts(filter: ModelPostFilterInput, limit: Int, nextToken: String): ModelPostConnection @aws_iam');
  const policyResources = _.filter(out.rootStack.Resources, r => r.Type === 'AWS::IAM::ManagedPolicy');
  expect(policyResources).toHaveLength(1);
  const resources = _.get(policyResources, '[0].Properties.PolicyDocument.Statement[0].Resource');
  const typeFieldList = _.map(resources, r => _.get(r, 'Fn::Sub[1]')).map(r => `${_.get(r, 'typeName')}.${_.get(r, 'fieldName', '*')}`);
  expect(typeFieldList).toEqual([
    'Post.*',
    'Query.getPost',
    'Query.listPosts',
    'Mutation.createPost',
    'Mutation.updatePost',
    'Mutation.deletePost',
    'Subscription.onCreatePost',
    'Subscription.onUpdatePost',
    'Subscription.onDeletePost',
  ]);
  // should throw unauthorized if it's not signed by the admin ui iam role
  ['Mutation.createPost.auth.1.req.vtl', 'Mutation.updatePost.auth.1.res.vtl', 'Mutation.deletePost.auth.1.res.vtl'].forEach(r => {
    expect(out.resolvers[r]).toMatchSnapshot();
  });
});

test('admin roles should be return the field name inside field resolvers', () => {
  const validSchema = `
    type Student @model @auth(rules: [{ allow: groups, groups: ["staff"] }, { allow: owner }]) {
      id: ID!
      name: String
      description: String
      secretValue: String @auth(rules: [{ allow: owner }])
    }`;
  const transformer = new GraphQLTransform({
    authConfig: {
      defaultAuthentication: {
        authenticationType: 'AMAZON_COGNITO_USER_POOLS',
      },
      additionalAuthenticationProviders: [
        {
          authenticationType: 'AWS_IAM',
        },
      ],
    },
    transformers: [
      new ModelTransformer(),
      new AuthTransformer({
        adminRoles: ADMIN_UI_ROLES,
      }),
    ],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();

  expect(out.resolvers['Student.secretValue.req.vtl']).toMatchSnapshot();
  expect(out.resolvers['Mutation.createStudent.auth.1.req.vtl']).toMatchSnapshot();
});
