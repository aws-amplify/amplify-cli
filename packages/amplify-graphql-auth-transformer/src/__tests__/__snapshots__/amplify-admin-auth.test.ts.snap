// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test simple model with AdminUI enabled should add IAM policy only for fields that have explicit IAM auth 1`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $inputFields = $util.parseJson($util.toJson($ctx.args.input.keySet())) )
#set( $isAuthorized = false )
#set( $allowedFields = [] )
#if( $util.authType() == \\"IAM Authorization\\" )
  #set( $adminRoles = [\\"us-fake-1_uuid_Full-access/CognitoIdentityCredentials\\",\\"us-fake-1_uuid_Manage-only/CognitoIdentityCredentials\\"] )
  #foreach( $adminRole in $adminRoles )
    #if( $ctx.identity.userArn.contains($adminRole) && $ctx.identity.userArn != $ctx.stash.authRole && $ctx.identity.userArn != $ctx.stash.unauthRole )
      #return($util.toJson({}))
    #end
  #end
$util.unauthorized()
#end
#if( $util.authType() == \\"User Pool Authorization\\" )

#end
#if( !$isAuthorized && $allowedFields.isEmpty() )
$util.unauthorized()
#end
#if( !$isAuthorized )
  #set( $deniedFields = $util.list.copyAndRemoveAll($inputFields, $allowedFields) )
  #if( $deniedFields.size() > 0 )
    $util.error(\\"Unauthorized on \${deniedFields}\\", \\"Unauthorized\\")
  #end
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`Test simple model with AdminUI enabled should add IAM policy only for fields that have explicit IAM auth 2`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#if( $ctx.error )
  $util.error($ctx.error.message, $ctx.error.type)
#end
#set( $inputFields = $util.parseJson($util.toJson($ctx.args.input.keySet())) )
#set( $isAuthorized = false )
#set( $allowedFields = [] )
#set( $nullAllowedFields = [] )
#set( $deniedFields = {} )
#if( $util.authType() == \\"IAM Authorization\\" )
  #set( $adminRoles = [\\"us-fake-1_uuid_Full-access/CognitoIdentityCredentials\\",\\"us-fake-1_uuid_Manage-only/CognitoIdentityCredentials\\"] )
  #foreach( $adminRole in $adminRoles )
    #if( $ctx.identity.userArn.contains($adminRole) && $ctx.identity.userArn != $ctx.stash.authRole && $ctx.identity.userArn != $ctx.stash.unauthRole )
      #return($util.toJson({}))
    #end
  #end
$util.unauthorized()
#end
#if( $util.authType() == \\"User Pool Authorization\\" )

#end
#if( !$isAuthorized && $allowedFields.isEmpty() && $nullAllowedFields.isEmpty() )
$util.unauthorized()
#end
#if( !$isAuthorized )
  #foreach( $entry in $util.map.copyAndRetainAllKeys($ctx.args.input, $inputFields).entrySet() )
    #if( $util.isNull($entry.value) && !$nullAllowedFields.contains($entry.key) )
      $util.qr($deniedFields.put($entry.key, \\"\\"))
    #end
  #end
  #foreach( $deniedField in $util.list.copyAndRemoveAll($inputFields, $allowedFields) )
    $util.qr($deniedFields.put($deniedField, \\"\\"))
  #end
#end
#if( $deniedFields.keySet().size() > 0 )
  $util.error(\\"Unauthorized on \${deniedFields.keySet()}\\", \\"Unauthorized\\")
#end
$util.toJson({})
## [End] Authorization Steps. **"
`;

exports[`Test simple model with AdminUI enabled should add IAM policy only for fields that have explicit IAM auth 3`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $isAuthorized = false )
#if( $util.authType() == \\"IAM Authorization\\" )
  #set( $adminRoles = [\\"us-fake-1_uuid_Full-access/CognitoIdentityCredentials\\",\\"us-fake-1_uuid_Manage-only/CognitoIdentityCredentials\\"] )
  #foreach( $adminRole in $adminRoles )
    #if( $ctx.identity.userArn.contains($adminRole) && $ctx.identity.userArn != $ctx.stash.authRole && $ctx.identity.userArn != $ctx.stash.unauthRole )
      #return($util.toJson({}))
    #end
  #end
$util.unauthorized()
#end
#if( $util.authType() == \\"User Pool Authorization\\" )

#end
#if( !$isAuthorized )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`Test simple model with private IAM auth rule, few operations, and amplify admin app is not enabled 1`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#if( $ctx.error )
  $util.error($ctx.error.message, $ctx.error.type)
#end
#set( $inputFields = $util.parseJson($util.toJson($ctx.args.input.keySet())) )
#set( $isAuthorized = false )
#set( $allowedFields = [] )
#set( $nullAllowedFields = [] )
#set( $deniedFields = {} )
#if( $util.authType() == \\"IAM Authorization\\" )
  #if( ($ctx.identity.userArn == $ctx.stash.authRole) || ($ctx.identity.cognitoIdentityPoolId == \\"testIdentityPoolId\\" && $ctx.identity.cognitoIdentityAuthType == \\"authenticated\\") )
    #set( $isAuthorized = true )
  #end
#end
#if( $util.authType() == \\"User Pool Authorization\\" )

#end
#if( !$isAuthorized && $allowedFields.isEmpty() && $nullAllowedFields.isEmpty() )
$util.unauthorized()
#end
#if( !$isAuthorized )
  #foreach( $entry in $util.map.copyAndRetainAllKeys($ctx.args.input, $inputFields).entrySet() )
    #if( $util.isNull($entry.value) && !$nullAllowedFields.contains($entry.key) )
      $util.qr($deniedFields.put($entry.key, \\"\\"))
    #end
  #end
  #foreach( $deniedField in $util.list.copyAndRemoveAll($inputFields, $allowedFields) )
    $util.qr($deniedFields.put($deniedField, \\"\\"))
  #end
#end
#if( $deniedFields.keySet().size() > 0 )
  $util.error(\\"Unauthorized on \${deniedFields.keySet()}\\", \\"Unauthorized\\")
#end
$util.toJson({})
## [End] Authorization Steps. **"
`;

exports[`admin roles should be return the field name inside field resolvers 1`] = `
"## [Start] Field Authorization Steps. **
#set( $isAuthorized = false )
#if( $util.authType() == \\"IAM Authorization\\" )
  #set( $adminRoles = [\\"us-fake-1_uuid_Full-access/CognitoIdentityCredentials\\",\\"us-fake-1_uuid_Manage-only/CognitoIdentityCredentials\\"] )
  #foreach( $adminRole in $adminRoles )
    #if( $ctx.identity.userArn.contains($adminRole) && $ctx.identity.userArn != $ctx.stash.authRole && $ctx.identity.userArn != $ctx.stash.unauthRole )
      #return($context.source.secretValue)
    #end
  #end
$util.unauthorized()
#end
#if( $util.authType() == \\"User Pool Authorization\\" )
  #if( !$isAuthorized )
    #set( $ownerEntity0 = $util.defaultIfNull($ctx.source.owner, null) )
    #set( $ownerClaim0 = $util.defaultIfNull($ctx.identity.claims.get(\\"username\\"), $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:username\\"), \\"___xamznone____\\")) )
    #if( $ownerEntity0 == $ownerClaim0 )
      #set( $isAuthorized = true )
    #end
  #end
#end
#if( !$isAuthorized )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Field Authorization Steps. **"
`;

exports[`admin roles should be return the field name inside field resolvers 2`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $inputFields = $util.parseJson($util.toJson($ctx.args.input.keySet())) )
#set( $isAuthorized = false )
#set( $allowedFields = [] )
#if( $util.authType() == \\"IAM Authorization\\" )
  #set( $adminRoles = [\\"us-fake-1_uuid_Full-access/CognitoIdentityCredentials\\",\\"us-fake-1_uuid_Manage-only/CognitoIdentityCredentials\\"] )
  #foreach( $adminRole in $adminRoles )
    #if( $ctx.identity.userArn.contains($adminRole) && $ctx.identity.userArn != $ctx.stash.authRole && $ctx.identity.userArn != $ctx.stash.unauthRole )
      #return($util.toJson({}))
    #end
  #end
$util.unauthorized()
#end
#if( $util.authType() == \\"User Pool Authorization\\" )
  #if( !$isAuthorized )
    #set( $staticGroupRoles = [{\\"claim\\":\\"cognito:groups\\",\\"entity\\":\\"staff\\",\\"allowedFields\\":[\\"id\\",\\"name\\",\\"description\\"]}] )
    #foreach( $groupRole in $staticGroupRoles )
      #set( $groupsInToken = $util.defaultIfNull($ctx.identity.claims.get($groupRole.claim), []) )
      #if( $groupsInToken.contains($groupRole.entity) )
        #if( !$groupRole.allowedFields.isEmpty() )
          $util.qr($allowedFields.addAll($groupRole.allowedFields))
        #else
          #set( $isAuthorized = true )
          #break
        #end
      #end
    #end
  #end
  #if( !$isAuthorized )
    #set( $ownerEntity0 = $util.defaultIfNull($ctx.args.input.owner, null) )
    #set( $ownerClaim0 = $util.defaultIfNull($ctx.identity.claims.get(\\"username\\"), $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:username\\"), \\"___xamznone____\\")) )
    #set( $ownerAllowedFields0 = [] )
    #if( $ownerClaim0 == $ownerEntity0 )
      #if( !$ownerAllowedFields0.isEmpty() )
        $util.qr($allowedFields.addAll($ownerAllowedFields0))
      #else
        #set( $isAuthorized = true )
      #end
    #end
    #if( $util.isNull($ownerEntity0) && !$ctx.args.input.containsKey(\\"owner\\") )
      $util.qr($ctx.args.input.put(\\"owner\\", $ownerClaim0))
      #if( !$ownerAllowedFields0.isEmpty() )
        $util.qr($allowedFields.addAll($ownerAllowedFields0))
      #else
        #set( $isAuthorized = true )
      #end
    #end
  #end
#end
#if( !$isAuthorized && $allowedFields.isEmpty() )
$util.unauthorized()
#end
#if( !$isAuthorized )
  #set( $deniedFields = $util.list.copyAndRemoveAll($inputFields, $allowedFields) )
  #if( $deniedFields.size() > 0 )
    $util.error(\\"Unauthorized on \${deniedFields}\\", \\"Unauthorized\\")
  #end
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;
